File: Sssl/SsslParser.cs
using System;
using System.Globalization;
using System.Runtime.Serialization;
using System.Text;

namespace Sssl
{
    internal class SsslParser
    {
        private readonly StringBuilder _builder = new StringBuilder();
        private readonly string _source;
        private int _index;

        public SsslParser(string source)
        {
            _source = source;
        }

        public SsslObject Parse()
        {
            SkipSpace();
            var result = GetValue();
            if (_index < _source.Length) throw Error("Unknown token.");
            return result;
        }

        private void Consume(char c)
        {
            _index++;
            _builder.Append(c);
        }

        private static bool IsDigit(char c) => '0' <= c && c <= '9';

        private static bool IsHex(char c) => '0' <= c && c <= '9' || 'a' <= c && c <= 'f' || 'A' <= c && c <= 'F';

        private static bool IsIdHead(char c) =>
            c == '_' || c == '$' || char.GetUnicodeCategory(c) switch
            {
                UnicodeCategory.UppercaseLetter => true,
                UnicodeCategory.LowercaseLetter => true,
                UnicodeCategory.TitlecaseLetter => true,
                UnicodeCategory.ModifierLetter => true,
                UnicodeCategory.OtherLetter => true,
                UnicodeCategory.LetterNumber => true,
                _ => false,
            };

        private static bool IsIdChar(char c) =>
            c == '_' || c == '$' || c == '.' || char.GetUnicodeCategory(c) switch
            {
                UnicodeCategory.UppercaseLetter => true,
                UnicodeCategory.LowercaseLetter => true,
                UnicodeCategory.TitlecaseLetter => true,
                UnicodeCategory.ModifierLetter => true,
                UnicodeCategory.OtherLetter => true,
                UnicodeCategory.LetterNumber => true,
                UnicodeCategory.NonSpacingMark => true,
                UnicodeCategory.SpacingCombiningMark => true,
                UnicodeCategory.DecimalDigitNumber => true,
                UnicodeCategory.ConnectorPunctuation => true,
                _ => false,
            };

        private char GetCurrent() => _index >= _source.Length ? ' ' : _source[_index];

        private void ConsumeWhile(Func<char, bool> predicate)
        {
            var c = GetCurrent();
            while (predicate(c))
            {
                Consume(c);
                c = GetCurrent();
            }
        }

        private void ConsumeWhileOne(Func<char, bool> predicate, string errorMessage)
        {
            var c = GetCurrent();
            if (!predicate(c)) throw Error(errorMessage);

            Consume(c);
            ConsumeWhile(predicate);
        }

        private void SkipSpace()
        {
            while (_index < _source.Length && " \t\r\n".Contains(GetCurrent()))
                _index++;
        }

        private SsslObject GetValue()
        {
            var c = GetCurrent();
            if (c == '(' || c == '{' || c == '[') return GetRecord("");
            
            if (IsDigit(c) || c == '-' || c == '+') return GetNumber();

            if (c == '"')
            {
                var str = GetString();
                return GetCurrent() switch
                {
                    ':' => (SsslObject)GetPair(str),
                    '(' => GetRecord(str),
                    '{' => GetRecord(str),
                    '[' => GetRecord(str),
                    _ => new SsslValue(str),
                };
            }
            
            if (IsIdHead(c))
            {
                var id = GetId();
                switch (id)
                {
                    case "true": return new SsslValue(true);
                    case "false": return new SsslValue(false);
                    case "null": return SsslValue.Null;
                    case "nan": return new SsslValue(double.NaN);
                    case "inf": return new SsslValue(double.PositiveInfinity);
                    case "ninf": return new SsslValue(double.NegativeInfinity);
                    default:
                        return GetCurrent() switch
                        {
                            ':' => (SsslObject)GetPair(id),
                            '(' => GetRecord(id),
                            '{' => GetRecord(id),
                            '[' => GetRecord(id),
                            _ => throw Error("Unknown token."),
                        };
                }
            }

            throw Error("Unknown token.");
        }

        private Exception Error(string message) =>
            new SsslParseException($"Invalid SSSL starting at character {_index},\r\n'{GetCurrent()}': {message}\r\n-----\r\n{_source[Math.Max(0, _index - 10)..Math.Min(_source.Length, _index + 10)]}", _source, _index);

        private SsslValue GetNumber()
        {
            var c = GetCurrent();
            _builder.Clear();

            if (c == '-' || c == '+')
            {
                Consume(c);
                c = GetCurrent();
            }

            if (c == '0') Consume(c);
            if ('1' <= c && c <= '9') ConsumeWhile(IsDigit);
            else throw Error("Digits are required.");

            // read fraction
            c = GetCurrent();
            if (c == '.')
            {
                Consume(c);
                ConsumeWhileOne(IsDigit, "Digits are required.");
            }

            // read exponent
            c = GetCurrent();
            if (c == 'e' || c == 'E')
            {
                Consume(c);
                c = GetCurrent();
                if (c == '-' || c == '+') Consume(c);
                ConsumeWhileOne(IsDigit, "Digits are required.");
            }

            SkipSpace();
            return new SsslValue(double.Parse(_builder.ToString()));
        }

        private string GetId()
        {
            var c = GetCurrent();
            _builder.Clear();

            if (!IsIdHead(c)) throw Error("A identifier is required.");

            Consume(c);
            ConsumeWhile(IsIdChar);
            SkipSpace();

            return _builder.ToString();
        }

        private string GetString()
        {
            var c = GetCurrent();
            _builder.Clear();

            if (c != '"') throw Error("A string is required.");

            _index++;
            c = GetCurrent();

            while (c != '"')
            {
                if (c == '\\')
                {
                    _index++;
                    c = GetCurrent();
                    switch (c)
                    {
                        case '\\': case '"': case '/': _builder.Append(c); break;
                        case 'b': _builder.Append('\b'); break;
                        case 'f': _builder.Append('\f'); break;
                        case 'n': _builder.Append('\n'); break;
                        case 'r': _builder.Append('\r'); break;
                        case 't': _builder.Append('\t'); break;
                        case 'u':
                            var codeChars = new char[4];
                            for (var i = 0; i < 4; i++)
                            {
                                _index++;
                                codeChars[i] = GetCurrent();
                                if (!IsHex(codeChars[i])) throw Error("Invalid escape sequence.");
                            }

                            _builder.Append((char)int.Parse(new string(codeChars), NumberStyles.HexNumber));
                            break;
                        default: throw Error("Invalid escape sequence.");
                    }
                }
                else
                {
                    _builder.Append(c);
                }

                _index++;
                c = GetCurrent();
            }

            _index++;
            SkipSpace();
            return _builder.ToString();
        }

        private SsslRecord GetRecord(string name)
        {
            var (endBracket, type) = GetCurrent() switch
            {
                '(' => (')', SsslRecordType.Parentheses),
                '{' => ('}', SsslRecordType.Braces),
                '[' => (']', SsslRecordType.Brackets),
                _ => throw Error("予期しないトークンです。"), // this exception is never thrown
            };

            _index++;
            SkipSpace();

            var result = new SsslRecord(type, name);
            while (true)
            {
                if (GetCurrent() == endBracket) break;
                result.Add(GetValue());
                var c = GetCurrent();
                if (c == endBracket) break;
                if (c != ',') throw Error("オブジェクトの終端が必要です。");
                _index++;
                SkipSpace();
            }

            _index++;
            SkipSpace();
            return result;
        }

        private SsslPair GetPair(string key)
        {
            if (GetCurrent() != ':') throw Error("予期しないトークンです。");
            _index++;
            SkipSpace();
            return new SsslPair(key, GetValue());
        }
    }

    [Serializable]
    public class SsslParseException : Exception
    {
        public string SourceText { get; }

        public int Index { get; }

        public override string Message => base.Message;

        public SsslParseException() => SourceText = "";

        public SsslParseException(string message, string sourceText, int index) : base(message)
        {
            SourceText = sourceText;
            Index = index;
        }

        public SsslParseException(string message) : base(message) => SourceText = "";

        public SsslParseException(string message, Exception inner) : base(message, inner) => SourceText = "";

        protected SsslParseException(SerializationInfo info, StreamingContext context) : base(info, context) => SourceText = "";
    }
}

File: SsslFSharp/SsslParser.fs
namespace SsslFSharp
open System
open System.Globalization
open System.Text

exception SsslParseException of message: string * sourceText: string * index: int with
    override this.Message =
        let beginIndex = max 0 (this.index - 10)
        let endIndex = min (this.sourceText.Length - 1) (this.index + 10)
        this.message + $"\r\nindex: %i{this.index}\r\n" +
        this.sourceText.[beginIndex..endIndex].Replace('\n', ' ').Replace('\r', ' ').Replace('\t', ' ') +
        "\r\n" + String(' ', this.index - beginIndex) + "^"


type internal SsslParser(source: string) =
    let builder = StringBuilder()
    let mutable index = 0

    static let isDigit c = '0' <= c && c <= '9'

    static let isHex c = '0' <= c && c <= '9' || 'a' <= c && c <= 'f' || 'A' <= c && c <= 'F'

    static let isIdHead c =
        c = '_' || c = '$' ||
        match Char.GetUnicodeCategory(c) with
        | UnicodeCategory.UppercaseLetter
        | UnicodeCategory.LowercaseLetter
        | UnicodeCategory.TitlecaseLetter
        | UnicodeCategory.ModifierLetter
        | UnicodeCategory.OtherLetter
        | UnicodeCategory.LetterNumber -> true
        | _ -> false

    static let isIdChar c =
        c = '_' || c = '$' || c = '.' ||
        match Char.GetUnicodeCategory(c) with
        | UnicodeCategory.UppercaseLetter
        | UnicodeCategory.LowercaseLetter
        | UnicodeCategory.TitlecaseLetter
        | UnicodeCategory.ModifierLetter
        | UnicodeCategory.OtherLetter
        | UnicodeCategory.LetterNumber
        | UnicodeCategory.NonSpacingMark
        | UnicodeCategory.SpacingCombiningMark
        | UnicodeCategory.DecimalDigitNumber
        | UnicodeCategory.ConnectorPunctuation -> true
        | _ -> false

    let error message = raise (SsslParseException(message, source, index))

    let consume (c: char) =
        index <- index + 1
        ignore <| builder.Append(c)

    let getCurrent() = if index >= source.Length then ' ' else source.[index]

    let consumeWhile predicate =
        let mutable c = getCurrent()
        while predicate c do
            consume c
            c <- getCurrent()

    let consumeWhileOne predicate errorMessage =
        let c = getCurrent()
        if not (predicate c) then error errorMessage
        consume c
        consumeWhile predicate

    let skipSpace() =
        while (index < source.Length && " \t\r\n".Contains(getCurrent())) do
            index <- index + 1

    member private _.GetNumber() =
        let mutable c = getCurrent()
        ignore <| builder.Clear()
        if c = '-' || c = '+' then
            consume c
            c <- getCurrent()
        if c = '0' then consume c
        elif '1' <= c && c <= '9' then consumeWhile isDigit
        else error "Digits are required."
        // read fraction
        c <- getCurrent()
        if c = '.' then
            consume c
            consumeWhileOne isDigit "Digits are required."
        // read exponent
        c <- getCurrent()
        if c = 'e' || c = 'E' then
            consume c
            c <- getCurrent()
            if c = '-' || c = '+' then consume c
            consumeWhileOne isDigit "Digits are required."
        skipSpace()
        Sssl.Number(Double.Parse(builder.ToString()))

    member private _.GetId() =
        let c = getCurrent()
        ignore <| builder.Clear()
        if not (isIdHead c) then error "A identifier is required."
        consume c
        consumeWhile isIdChar
        skipSpace()
        builder.ToString()

    member private _.GetString() =
        let mutable c = getCurrent()
        ignore <| builder.Clear()
        if c <> '"' then error "A string is required."
        index <- index + 1
        c <- getCurrent()
        while c <> '"' do
            if c = '\\' then
                index <- index + 1
                c <- getCurrent()
                match c with
                | '\\' | '"' | '/' -> c
                | 'b' -> '\b'
                | 'f' -> '\f'
                | 'n' -> '\n'
                | 'r' -> '\r'
                | 't' -> '\t'
                | 'u' ->
                    let codeChars = Array.zeroCreate 4
                    for i in 0..3 do
                        index <- index + 1
                        codeChars.[i] <- getCurrent()
                        if not (isHex codeChars.[i]) then error "Invalid escape sequence."
                    char (Int32.Parse(String(codeChars), NumberStyles.HexNumber))
                | _ -> error "Invalid escape sequence."
            else c
            |> builder.Append |> ignore
            index <- index + 1
            if index >= source.Length then error "End of string end not found."
            c <- getCurrent()
        index <- index + 1
        skipSpace()
        builder.ToString()

    member private this.GetPair(key) =
        if getCurrent() <> ':' then error "Unexpected token"
        index <- index + 1
        skipSpace()
        Sssl.Pair(key, this.GetValue())

    member private this.GetRecord(name) =
        let endBracket, recordType =
            match getCurrent() with
            | '(' -> ')', SsslRecordType.Parentheses
            | '{' -> '}', SsslRecordType.Braces
            | '[' -> ']', SsslRecordType.Brackets
            | _ -> error "予期しないトークンです。" // this exception is never thrown
        index <- index + 1
        skipSpace()
        let contents = ResizeArray()
        let rec readContent() =
            if getCurrent() = endBracket then ()
            else
                contents.Add(this.GetValue())
                let c = getCurrent()
                if c = endBracket then ()
                elif c <> ',' then error "オブジェクトの終端が必要です。"
                else
                    index <- index + 1
                    skipSpace()
                    readContent()
        readContent()
        index <- index + 1
        skipSpace()
        Sssl.Record(name, recordType, contents.ToArray())

    member private this.GetValue() =
        match getCurrent() with
        | '(' | '{' | '[' -> this.GetRecord("")
        | '"' ->
            let str = this.GetString()
            match getCurrent() with
            | ':' -> this.GetPair(str)
            | '(' | '{' | '[' -> this.GetRecord(str)
            | _ -> Sssl.String(str)
        | '-' | '+' -> this.GetNumber()
        | c when isDigit c -> this.GetNumber()
        | c when isIdHead c ->
            match this.GetId() with
            | "true" -> Sssl.Bool(true)
            | "false" -> Sssl.Bool(false)
            | "null" -> Sssl.Null
            | "nan" -> Sssl.Number(nan)
            | "inf" -> Sssl.Number(infinity)
            | "ninf" -> Sssl.Number(-infinity)
            | name ->
                match getCurrent() with
                | ':' -> this.GetPair(name)
                | '(' | '{' | '[' -> this.GetRecord(name)
                | _ -> error "Unknown token in literal."
        | _ -> error "Unknown token in value."

    member this.Parse() =
        skipSpace()
        let result = this.GetValue()
        if index < source.Length then error "Unknown token."
        result
